functorch/functorch/experimental/cond.py

```diff
@@ -0,0 +1,158 @@
+        # evalauted as a torch.Tensor.
+        # So we execute the real true and false fn here and compare metadata
+        true_result = true_fn(*operands)
+        false_result = false_fn(*operands)
+        def recursive_compare_same(a, b):
+            assert(type(a) == type(b))
+            if isinstance(a, torch.Tensor):
+                assert(a.dtype == b.dtype)
+                assert(a.size() == b.size())
+                assert(a.stride() == b.stride())
+                assert(a.device == b.device)
+            elif isinstance(a, (list, tuple)):
+                assert(len(a) == len(b))
+                for i in range(0, len(a)):
+                    recursive_compare_same(a[i], b[i])
+
+        recursive_compare_same(true_result, false_result)
+        
+    args = (pred, true_graph, false_graph, operands)
+
+    proxy_args = pytree.tree_map(_unwrap_proxy, args)
+
+    # Does this need random slug appended so as not to collide?
+    proxy_res = proxy_mode.tracer.create_proxy('call_function', func_overload, proxy_args, kwargs,
+                                            name="conditional")
+    return proxy_res
+
+
+def cond_dense(pred, true_fn, false_fn, *operands):
+    mode = torch._C._get_torch_dispatch_mode()
+    assert (mode is None), "Mode should never be enabled for CPU key"
+    try:
+        if pred:
+            return true_fn(*operands)
+        else:
+            return false_fn(*operands)
+    except Exception as e:
+        # Do something proper here, someday
+        print("Exception:", e)
+        raise e
+
+
+def cond_autograd(pred, true_fn, false_fn, *operands):
+    # TODO: support autograd
+    flat_operands, _ = tree_flatten((true_fn, false_fn) + operands)
+    assert all([not f.requires_grad for f in flat_operands
+                if isinstance(f, torch.Tensor)])
+
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.AutogradCPU))
+    return cond(pred, true_fn, false_fn, *operands)
+
+
+def cond_adinplaceorview(*args, **kwargs):
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.ADInplaceOrView))
+    return cond(*args, **kwargs)
+
+def cond_backendselect(*args, **kwargs):
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.BackendSelect))
+    return cond(*args, **kwargs)
+
+def python_fallback(op):
+    def inner(*args, **kwargs):
+        mode = torch._C._get_torch_dispatch_mode()
+        if mode:
+            with suspend_mode(mode):
+                res = trace_cond(mode, op, args, kwargs)
+            return res
+        else:
+            return cond(args, kwargs)
+    return inner
+
+
+cond = PyOperator('cond')
+cond.impl(DispatchKey.CPU, cond_dense)
+cond.impl(DispatchKey.Python, python_fallback(cond))
+cond.fallthrough(DispatchKey.PythonTLSSnapshot)
+cond.impl(DispatchKey.AutogradCPU, cond_autograd)
+cond.fallthrough(DispatchKey.ADInplaceOrView)
+cond.fallthrough(DispatchKey.BackendSelect)
```

Comment:
What Ed said above: it would be more faithful for these to look like `cond.impl(DispatchKey.ADInplaceOrView, fallthrough_fn)`

Reply:
sgtm